File: examples/canonical-auth/java/src/test/java/dev/tenantcore/auth/AuthIntegrationTest.java
package dev.tenantcore.auth;

import io.jsonwebtoken.Jwts;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import dev.tenantcore.auth.auth.JwtService;
import dev.tenantcore.auth.domain.InMemoryStore;
import dev.tenantcore.auth.domain.User;

import javax.crypto.SecretKey;
import java.util.Date;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
public class AuthIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private JwtService jwtService;

    @Autowired
    private InMemoryStore store;

    @Autowired
    private SecretKey jwtSigningKey;

    @Test
    public void token_issue_success() throws Exception {
        mockMvc.perform(post("/auth/token")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"email\":\"admin@acme.example\", \"password\":\"password\"}"))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.access_token").exists())
            .andExpect(jsonPath("$.token_type").value("Bearer"));
    }

    @Test
    public void token_invalid_credentials() throws Exception {
        mockMvc.perform(post("/auth/token")
                .contentType(MediaType.APPLICATION_JSON)
                .content("{\"email\":\"admin@acme.example\", \"password\":\"wrong\"}"))
            .andExpect(status().isUnauthorized());
    }

    @Test
    public void token_expired_rejection() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        Date past = new Date(System.currentTimeMillis() - 100000);
        String token = Jwts.builder()
            .subject(user.id())
            .issuer("tenantcore-auth")
            .claim("tenant_id", user.tenantId())
            .claim("acl_ver", user.aclVer())
            .expiration(past)
            .signWith(jwtSigningKey)
            .compact();

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isUnauthorized());
    }

    @Test
    public void token_stale_acl_ver() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        String token = Jwts.builder()
            .subject(user.id())
            .issuer("tenantcore-auth")
            .claim("tenant_id", user.tenantId())
            .claim("acl_ver", user.aclVer() - 1) // stale
            .expiration(new Date(System.currentTimeMillis() + 100000))
            .signWith(jwtSigningKey)
            .compact();

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isUnauthorized());
    }

    @Test
    public void get_me_success() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.sub").value(user.id()));
    }

    @Test
    public void get_me_unauthorized() throws Exception {
        mockMvc.perform(get("/me"))
            .andExpect(status().isUnauthorized());
    }

    @Test
    public void rbac_permission_granted() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/users").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk());
    }

    @Test
    public void rbac_permission_denied() throws Exception {
        User user = store.getUserByEmail("bob@acme.example");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(post("/groups")
                .header("Authorization", "Bearer " + token)
                .contentType(MediaType.APPLICATION_JSON)
                .content("{}"))
            .andExpect(status().isForbidden());
    }

    @Test
    public void cross_tenant_denial() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        String token = jwtService.issueToken(user, store);
        
        // Target a user in a different tenant (Globex)
        User target = store.getUserByEmail("admin@globex.example");

        mockMvc.perform(get("/users/" + target.id())
                .header("Authorization", "Bearer " + token))
            .andExpect(status().isForbidden());
    }

    @Test
    public void super_admin_access() throws Exception {
        User user = store.getUserByEmail("superadmin@tenantcore.dev");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/admin/tenants")
                .header("Authorization", "Bearer " + token))
            .andExpect(status().isOk());
    }

    @Test
    public void super_admin_tenant_scoped_denial() throws Exception {
        User user = store.getUserByEmail("superadmin@tenantcore.dev");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/users").header("Authorization", "Bearer " + token))
            .andExpect(status().isForbidden());
    }

    @Test
    public void me_allowed_routes_match_permissions() throws Exception {
        User user = store.getUserByEmail("admin@acme.example");
        String token = jwtService.issueToken(user, store);

        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.allowed_routes[?(@.path == '/users')]").exists());
            
        // Test that a route not allowed is omitted
        mockMvc.perform(get("/me").header("Authorization", "Bearer " + token))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.allowed_routes[?(@.path == '/admin/tenants')]").doesNotExist());
    }
}

File: examples/canonical-auth/kotlin/src/test/kotlin/dev/tenantcore/auth/AuthIntegrationTest.kt
package dev.tenantcore.auth

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import dev.tenantcore.auth.auth.TokenService
import dev.tenantcore.auth.domain.InMemoryStore
import io.jsonwebtoken.Jwts
import org.http4k.core.Method.GET
import org.http4k.core.Method.PATCH
import org.http4k.core.Method.POST
import org.http4k.core.Request
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import java.nio.file.Path
import java.util.Date

class AuthIntegrationTest {
    private val store = InMemoryStore.loadFromFixtures(Path.of("..", "domain", "fixtures"))
    private val tokenService = TokenService()
    private val handler = app(store, tokenService)
    private val mapper = jacksonObjectMapper()

    @Test
    fun token_issue_success() {
        val response = handler(
            Request(POST, "/auth/token")
                .header("content-type", "application/json")
                .body("""{"email":"admin@acme.example","password":"password"}"""),
        )
        assertEquals(200, response.status.code)
        val payload = mapper.readValue(response.bodyString(), Map::class.java)
        assertTrue(payload.containsKey("access_token"))
    }

    @Test
    fun token_invalid_credentials() {
        val response = handler(
            Request(POST, "/auth/token")
                .header("content-type", "application/json")
                .body("""{"email":"admin@acme.example","password":"wrong"}"""),
        )
        assertEquals(401, response.status.code)
    }

    @Test
    fun token_expired_rejection() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val token = Jwts.builder()
            .subject(user.id)
            .issuer(TokenService.ISSUER)
            .audience().add(TokenService.AUDIENCE).and()
            .claim("tenant_id", user.tenantId)
            .claim("tenant_role", "tenant_admin")
            .claim("groups", listOf("iam-admins"))
            .claim("permissions", store.getEffectivePermissions(user.id))
            .claim("acl_ver", user.aclVer)
            .issuedAt(Date(System.currentTimeMillis() - 2_000))
            .notBefore(Date(System.currentTimeMillis() - 2_000))
            .expiration(Date(System.currentTimeMillis() - 1_000))
            .signWith(TokenService.defaultSigningKey())
            .compact()

        val response = handler(Request(GET, "/me").header("Authorization", "Bearer $token"))
        assertEquals(401, response.status.code)
    }

    @Test
    fun token_stale_acl_ver() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val token = Jwts.builder()
            .subject(user.id)
            .issuer(TokenService.ISSUER)
            .audience().add(TokenService.AUDIENCE).and()
            .claim("tenant_id", user.tenantId)
            .claim("tenant_role", "tenant_admin")
            .claim("groups", listOf("iam-admins"))
            .claim("permissions", store.getEffectivePermissions(user.id))
            .claim("acl_ver", user.aclVer - 1)
            .issuedAt(Date())
            .notBefore(Date())
            .expiration(Date(System.currentTimeMillis() + 10_000))
            .signWith(TokenService.defaultSigningKey())
            .compact()

        val response = handler(Request(GET, "/me").header("Authorization", "Bearer $token"))
        assertEquals(401, response.status.code)
    }

    @Test
    fun get_me_success() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val token = tokenService.issueToken(user, store)
        val response = handler(Request(GET, "/me").header("Authorization", "Bearer $token"))
        assertEquals(200, response.status.code)
    }

    @Test
    fun get_me_unauthorized() {
        val response = handler(Request(GET, "/me"))
        assertEquals(401, response.status.code)
    }

    @Test
    fun rbac_permission_granted() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val token = tokenService.issueToken(user, store)
        val response = handler(Request(GET, "/users").header("Authorization", "Bearer $token"))
        assertEquals(200, response.status.code)
    }

    @Test
    fun rbac_permission_denied() {
        val user = checkNotNull(store.getUserByEmail("bob@acme.example"))
        val token = tokenService.issueToken(user, store)
        val response = handler(
            Request(POST, "/groups")
                .header("Authorization", "Bearer $token")
                .header("content-type", "application/json")
                .body("""{"slug":"extra","name":"Extra"}"""),
        )
        assertEquals(403, response.status.code)
    }

    @Test
    fun cross_tenant_denial() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val target = checkNotNull(store.getUserByEmail("admin@globex.example"))
        val token = tokenService.issueToken(user, store)
        val response = handler(Request(GET, "/users/${target.id}").header("Authorization", "Bearer $token"))
        assertEquals(403, response.status.code)
    }

    @Test
    fun super_admin_access() {
        val user = checkNotNull(store.getUserByEmail("superadmin@tenantcore.dev"))
        val token = tokenService.issueToken(user, store)
        val response = handler(Request(GET, "/admin/tenants").header("Authorization", "Bearer $token"))
        assertEquals(200, response.status.code)
    }

    @Test
    fun super_admin_tenant_scoped_denial() {
        val user = checkNotNull(store.getUserByEmail("superadmin@tenantcore.dev"))
        val token = tokenService.issueToken(user, store)
        val response = handler(Request(GET, "/users").header("Authorization", "Bearer $token"))
        assertEquals(403, response.status.code)
    }

    @Test
    fun me_allowed_routes_match_permissions() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val token = tokenService.issueToken(user, store)
        val response = handler(Request(GET, "/me").header("Authorization", "Bearer $token"))
        val payload = mapper.readValue(response.bodyString(), Map::class.java)
        val allowedRoutes = payload["allowed_routes"] as List<Map<String, Any?>>
        assertTrue(allowedRoutes.any { it["path"] == "/users" })
        assertFalse(allowedRoutes.any { it["path"] == "/admin/tenants" })
    }
}
